### ISSUE TITLE
Differential Backups of ceph image not always not working

### ISSUE THREAD

**Issue by Gert-P** [elemental-lf/benji#53]
Following the steps on https://benji-backup.me/backup.html#examples to do differential backups works only some times, but most of the times I encounter the following problem:

> (benji) root@node3:/# rbd snap create ceph-ssd/vm-111-disk-1@backup1
> (benji) root@node3:/# rbd diff --whole-object ceph-ssd/vm-111-disk-1@backup1 --format=json > /tmp/vm1.diff
> (benji) root@node3:/# benji backup --snapshot-name backup1 --rbd-hints /tmp/vm1.diff rbd:ceph-ssd/vm-111-disk-1@backup1 vm1
>     INFO: $ /usr/local/benji/bin/benji backup --snapshot-name backup1 --rbd-hints /tmp/vm1.diff rbd:ceph-ssd/vm-111-disk-1@backup1 vm1
>     INFO: Backed up 1/15 blocks (6.7%)
>     INFO: Backed up 2/15 blocks (13.3%)
>     INFO: Backed up 3/15 blocks (20.0%)
>     INFO: Backed up 4/15 blocks (26.7%)
>     INFO: Backed up 5/15 blocks (33.3%)
>     INFO: Backed up 6/15 blocks (40.0%)
>     INFO: Backed up 7/15 blocks (46.7%)
>     INFO: Backed up 8/15 blocks (53.3%)
>     INFO: Backed up 9/15 blocks (60.0%)
>     INFO: Backed up 10/15 blocks (66.7%)
>     INFO: Backed up 11/15 blocks (73.3%)
>     INFO: Backed up 12/15 blocks (80.0%)
>     INFO: Backed up 13/15 blocks (86.7%)
>     INFO: Backed up 14/15 blocks (93.3%)
>     INFO: Backed up 15/15 blocks (100.0%)
>     INFO: Set status of version V0000000001 to valid.
>     INFO: Backed up metadata of version V0000000001.
>     INFO: New version V0000000001 created, backup successful.

> (benji) root@node3:/# rbd snap create ceph-ssd/vm-111-disk-1@backup2
> (benji) root@node3:/# rbd diff --whole-object ceph-ssd/vm-111-disk-1@backup2 --from-snap backup1 --format=json > /tmp/vm1.diff
> (benji) root@node3:/# benji ls 'name == "vm1" and snapshot_name == "backup1"'
>     INFO: $ /usr/local/benji/bin/benji ls name == "vm1" and snapshot_name == "backup1"
> +---------------------+-------------+------+---------------+--------+------------+--------+-----------+---------+
> |         date        |     uid     | name | snapshot_name |   size | block_size | status | protected | storage |
> +---------------------+-------------+------+---------------+--------+------------+--------+-----------+---------+
> | 2019-09-03T02:59:44 | V0000000001 | vm1  | backup1       | 1.0GiB |     4.0MiB | valid  |   False   | usb     |
> +---------------------+-------------+------+---------------+--------+------------+--------+-----------+---------+
> (benji) root@node3:/# benji backup --snapshot-name backup2 --rbd-hints /tmp/vm1.diff --base-version V0000000001 rbd:ceph-ssd/vm-111-disk-1@backup2 vm1
>     INFO: $ /usr/local/benji/bin/benji backup --snapshot-name backup2 --rbd-hints /tmp/vm1.diff --base-version V0000000001 rbd:ceph-ssd/vm-111-disk-1@backup2 vm1
>  WARNING: Hints are empty, assuming nothing has changed.
>     INFO: Starting sanity check with 0.1% of the ignored blocks.
>    ERROR: Source and backup don't match in regions outside of the ones indicated by the hints.
>    ERROR: Looks like the hints don't match or the source is different.
>    ERROR: Found wrong source data at block 4: offset 16777216, length 4194304
>    ERROR: InputDataError: Source changed in regions outside of ones indicated by the hints.

**Comment by elemental-lf** [comment]
I haven't encountered this error apart from tests and human error but your calling sequence looks correct to me. In your second backup the hints are empty (`WARNING: Hints are empty, assuming nothing has changed.`), indicating that nothing has changed. But in the sanity check Benji finds differences between the first backup and the second snapshot where there should be none. 

Can you say if there should be differences between the first and second snapshot?
What is the contents of the second RBD hints file?
Does this only happen when there are no differences?
Could you please try to deep-scrub V1 to see if there is anything wrong with the first backup?
Could you please try to compare V1 with the first snapshot by using the `--source` option to `deep-scrub`?
What Ceph version are you using? On the client-side and the version of the cluster.
What is the version of Benji?
What is Benji's configuration? (Remember to redact any password, access keys and secrets.)
Are you using the `rbd` or the `rbdaio` i/o module to access the Ceph cluster?
Could you please try to use the other i/o module and see if it makes a difference?

**Comment by Gert-P** [comment]
> Can you say if there should be differences between the first and second snapshot?

As this was just a test there was no changes and the empty hints would be correct, however I also tried after copying random 250MB data to the image and the results were the same, interestingly enough the hints files was also empty.



> What is the contents of the second RBD hints file?

[]


> Does this only happen when there are no differences?

No, it makes no difference whether the image changed or not. Also the hints file is empty even if the image changed


> Could you please try to deep-scrub V1 to see if there is anything wrong with the first backup?
Could you please try to compare V1 with the first snapshot by using the --source option to deep-scrub?

Deep scrub is successful

```
(benji) root@node3:~# benji deep-scrub -s rbd:ceph-ssd/vm-111-disk-1@backup1 1
    INFO: $ /usr/local/benji/bin/benji deep-scrub -s rbd:ceph-ssd/vm-111-disk-1@backup1 1
    INFO: Deep-scrubbed 1/8 blocks (12.5%)
    INFO: Deep-scrubbed 2/8 blocks (25.0%)
    INFO: Deep-scrubbed 3/8 blocks (37.5%)
    INFO: Deep-scrubbed 4/8 blocks (50.0%)
    INFO: Deep-scrubbed 5/8 blocks (62.5%)
    INFO: Deep-scrubbed 6/8 blocks (75.0%)
    INFO: Deep-scrubbed 7/8 blocks (87.5%)
    INFO: Deep-scrubbed 8/8 blocks (100.0%)
    INFO: Set status of version V0000000001 to valid.
    INFO: Deep-scrub of version V0000000001 successful.
```



> What Ceph version are you using? On the client-side and the version of the cluster.

```
(benji) root@node3:/# ceph -v
ceph version 14.2.2 (a887fe9a5d3d97fe349065d3c1c9dbd7b8870855) nautilus (stable)

```


> What is the version of Benji?

```
(benji) root@node3:/# benji version-info
    INFO: $ /usr/local/benji/bin/benji version-info
    INFO: Benji version: 0.7.0.dev11+g51b7886.
    INFO: Configuration version: 1, supported >=1,<2.
    INFO: Metadata version: 1.1.0, supported >=1,<2.
    INFO: Object metadata version: 1.0.0, supported >=1,<2.

```


> What is Benji's configuration? (Remember to redact any password, access keys and secrets.)

```
configurationVersion: '1'
logFile: /var/log/benji.log
databaseEngine: sqlite:////var/lib/benji/benji.sqlite
blockSize: 4194304
processName: benji
disallowRemoveWhenYounger: 6
defaultStorage: usb
storages:
  - name: usb
    storageId: 1
    module: file
    configuration:
      path: /mnt/usb/benji-data/
  - name: b2
    storageId: 2
    module: b2
    configuration:
      accountId: ******
      applicationKey: ******
      bucketName: ******
      writeObjectAttempts: 2
      uploadAttempts: 5
      bandwidthRead: 0
      bandwidthWrite: 0
      simultaneousWrites: 20
      simultaneousReads: 20
ios:
  - name: rbd
    module: rbd
    configuration:
      cephConfigFile: /etc/ceph/ceph.conf
      clientIdentifier: admin
      newImageFeatures:
        - RBD_FEATURE_LAYERING
        - RBD_FEATURE_EXCLUSIVE_LOCK
        - RBD_FEATURE_STRIPINGV2
      simultaneousReads: 20
  - name: file
    module: file

```


> Are you using the rbd or the rbdaio i/o module to access the Ceph cluster?
Could you please try to use the other i/o module and see if it makes a difference?

I was using rbd but I also changed it to rbdaio and the results were identical.

**Comment by Gert-P** [comment]
Maybe the problem is not benji related, it seems like everytime --from-snap is used with the "rbd diff" the hints are empty regardless if anything changed in the image or not. See this:


> (benji) root@node3:/# rbd diff --whole-object ceph-ssd/vm-111-disk-1@backup2 --from-snap backup1 --format=json
> []
> (benji) root@node3:/# rbd diff --whole-object ceph-ssd/vm-111-disk-1@backup2  --format=json
> [{"offset":0,"length":4194304,"exists":"true"},{"offset":16777216,"length":4194304,"exists":"true"},{"offset":134217728,"length":4194304,"exists":"true"},{"offset":138412032,"length":4194304,"exists":"true"},{"offset":402653184,"length":4194304,"exists":"true"},{"offset":536870912,"length":4194304,"exists":"true"},{"offset":541065216,"length":4194304,"exists":"true"},{"offset":545259520,"length":4194304,"exists":"true"},{"offset":549453824,"length":4194304,"exists":"true"},{"offset":553648128,"length":4194304,"exists":"true"},{"offset":557842432,"length":4194304,"exists":"true"},{"offset":562036736,"length":4194304,"exists":"true"},{"offset":566231040,"length":4194304,"exists":"true"},{"offset":671088640,"length":4194304,"exists":"true"},{"offset":939524096,"length":4194304,"exists":"true"},{"offset":1069547520,"length":4194304,"exists":"true"}]
>

**Comment by Gert-P** [comment]
It seems that the problem only occurs with lxc container images, as kvm images works as expected.

**Comment by elemental-lf** [comment]
Thanks for the info. You say that the hints are empty even when there were changes between snapshots. As the hints are generated by `rbd diff` Benji isn't the culprit here. But I'd still be interested in finding out what is wrong here. A quick search in the Ceph issue tracker yielded nothing. I have the feeling we're missing something significant.
> It seems that the problem only occurs with lxc container images, as kvm images works as expected.

This seems strange to me. Is there any difference between these images? Do they have different image features? Is anything from Proxmox interfering  with the RBD snapshots in the lxc case?
(As a stop-gap measure you can try if everything works okay if you don't supply any hints at all. This will be somewhat ineffective as the snapshot would always be read in full but the blocks would still be deduplicated.)

**Comment by Gert-P** [comment]
Features are not identical:

```
rbd image 'vm-111-disk-0':
        size 5 GiB in 1280 objects
        order 22 (4 MiB objects)
        snapshot_count: 1
        id: 227fa1daeb8fe
        block_name_prefix: rbd_data.227fa1daeb8fe
        format: 2
        features: layering, exclusive-lock, object-map, fast-diff
        op_features: 
        flags: 
        create_timestamp: Sat Aug 17 08:28:52 2019
        access_timestamp: Tue Sep  3 18:49:37 2019
        modify_timestamp: Sat Aug 17 08:28:52 2019
root@node1:~# rbd info ceph-ssd/vm-106-disk-0
rbd image 'vm-106-disk-0':
        size 32 GiB in 8192 objects
        order 22 (4 MiB objects)
        snapshot_count: 1
        id: 571a5ebbe00ad
        block_name_prefix: rbd_data.571a5ebbe00ad
        format: 2
        features: layering, exclusive-lock, object-map, fast-diff, deep-flatten
        op_features: 
        flags: 
        create_timestamp: Sat Aug 17 10:15:40 2019
        access_timestamp: Tue Sep  3 23:25:01 2019
        modify_timestamp: Tue Sep  3 23:37:34 2019
root@node1:~# 
```
it works ok without hints. 

Here is something interesting I noticed:

First we do a test with a KVM image:
```
root@node1:~# rbd du --pool ceph-ssd | grep 106
vm-106-disk-0                   32 GiB 8.0 GiB 
root@node1:~# rbd snap create ceph-ssd/vm-106-disk-0@backup1
root@node1:~# rbd du --pool ceph-ssd | grep 106
vm-106-disk-0@backup1           32 GiB 8.0 GiB 
vm-106-disk-0                   32 GiB     0 B 
root@node1:~# ssh 11.2.0.15
root@11.2.0.15's password: 
Last login: Tue Sep  3 18:56:50 2019
[root@smtprelay ~]# dd if=/dev/urandom of=/temp/testfile bs=1M count=1024
1024+0 records in
1024+0 records out
1073741824 bytes (1.1 GB) copied, 73.4063 s, 14.6 MB/s
[root@smtprelay ~]# exit
root@node1:~# rbd du --pool ceph-ssd | grep 106
vm-106-disk-0@backup1           32 GiB 8.0 GiB 
vm-106-disk-0                   32 GiB 1.1 GiB 
```
As you can see the snapshot image size increase from 0B to 1.1GB after adding 1.1GB of random data to it, now we do the same with the LXC image:

```
root@node1:~# rbd snap create ceph-ssd/vm-111-disk-0@backup1
root@node1:~# rbd du --pool ceph-ssd | grep 111
vm-111-disk-0@backup1            5 GiB 2.3 GiB 
vm-111-disk-0                    5 GiB     0 B 
root@node1:~# lxc-attach --name 111
[root@remdesk ~]# dd if=/dev/urandom of=/temp/testfile bs=1M count=1024
1024+0 records in
1024+0 records out
1073741824 bytes (1.1 GB) copied, 9.24108 s, 116 MB/s
[root@remdesk ~]# exit
root@node1:~# rbd du --pool ceph-ssd | grep 111
vm-111-disk-0@backup1            5 GiB 2.3 GiB 
vm-111-disk-0                    5 GiB     0 B 
```

And here you can see that according to "rbd du" the size has not increased at all and is still showing 0B

Then after i rebuild the object-map the usage was reported correctly

```
root@node1:~# rbd object-map rebuild ceph-ssd/vm-111-disk-0
Object Map Rebuild: 100% complete...done.
root@node1:~# rbd du --pool ceph-ssd | grep 111
vm-111-disk-0@backup1            5 GiB 2.3 GiB 
vm-111-disk-0                    5 GiB 964 MiB 
```

So I tried the differential backup again, but this time rebuilding the object map after creating the snapshot, and the backup was successful. It looks like the feature "fast-diff" might be the culprit here, although it does not cause the same problem on KVM images. I'm not sure if this is normal of a bug?  I will test again with "fast-diff" disabled and see if that works.

**Comment by Gert-P** [comment]
As I suspected it is working with fast-diff feature disabled on the lxc image, so fast-diff on lxc images is indeed the culprit here. do you think it is a bug?

**Comment by elemental-lf** [comment]
This might be related to https://github.com/elemental-lf/benji/issues/14 which @olifre reported and which resulted in https://tracker.ceph.com/issues/37876 due to Oliver's research. But he didn't see any data corruption.
The thing is that at least on the Kubernetes side of things most people will be using the KRBD driver which as at Jewel feature level at least in Linux 4.19. Furthermore the (old) dynamic provisioners only support `layering` and nothing more. So nobody would see such a bug there. Which might explain why it hasn't been investigated yet.

**Comment by elemental-lf** [comment]
I just found https://tracker.ceph.com/issues/39435 which sounds like it fixes the problem Oliver had with inconsistent omaps. This fix and at least one other RBD related fix will be in 14.2.3. So it probably makes sense to test with 14.2.3 when it comes out shortly.

**Comment by Gert-P** [comment]
Great, I hope so, for now I am just rebuilding the object map before creating the hints file and it is working great. Thank you once again for your interest in the problem and prompt response

**Comment by olifre** [comment]
> But he didn't see any data corruption.

@elemental-lf Just to complete the picture, we also used KVM, and for us, the object map differences were caused by trim / DISCARD. 
We will not migrate to 14.2.3 quickly (our production virtualization cluster is still on Mimic), but of course we'll do it at some point after gaining more experience with Nautilus. So many thanks for pointing out the tracker bug to me, I'll report back here once we use a fixed release and can confirm the fix / know if the problem persists.
